File: go-vs-python-data-processing/python/process_data.py
import pandas as pd
import time
import psutil
import os
from pathlib import Path

class DataProcessor:
    """
    Classe para processar dados e medir performance
    """
    
    def __init__(self):
        self.process = psutil.Process()
        self.results = {}
    
    def measure_resources(self):
        """Mede uso atual de CPU e memória"""
        return {
            'cpu_percent': self.process.cpu_percent(),
            'memory_mb': self.process.memory_info().rss / 1024 / 1024,
            'memory_percent': self.process.memory_percent()
        }
    
    def read_csv_with_timing(self, filepath, method='pandas'):
        """
        Lê CSV e mede tempo de execução e recursos
        """
        print(f"\n{'='*50}")
        print(f"LENDO CSV: {os.path.basename(filepath)}")
        print(f"Método: {method}")
        print(f"{'='*50}")
        
        # Verificar se arquivo existe
        if not os.path.exists(filepath):
            print(f"❌ Arquivo não encontrado: {filepath}")
            return None
        
        # Informações do arquivo
        file_size = os.path.getsize(filepath)
        print(f"📁 Tamanho do arquivo: {file_size:,} bytes ({file_size/1024/1024:.2f} MB)")
        
        # Recursos iniciais
        resources_before = self.measure_resources()
        print(f"🔋 Recursos ANTES - CPU: {resources_before['cpu_percent']:.1f}% | "
              f"Memória: {resources_before['memory_mb']:.1f} MB ({resources_before['memory_percent']:.1f}%)")
        
        # Medir tempo de leitura
        start_time = time.time()
        
        try:
            if method == 'pandas':
                df = pd.read_csv(filepath)
            else:
                # Método alternativo usando csv padrão
                import csv
                data = []
                with open(filepath, 'r', encoding='utf-8') as file:
                    reader = csv.DictReader(file)
                    data = list(reader)
                df = pd.DataFrame(data)
                df['value'] = pd.to_numeric(df['value'])
            
            end_time = time.time()
            execution_time = end_time - start_time
            
            # Recursos finais
            resources_after = self.measure_resources()
            
            # Informações do dataset
            print(f"\n📊 DADOS CARREGADOS:")
            print(f"   • Linhas: {len(df):,}")
            print(f"   • Colunas: {len(df.columns)}")
            print(f"   • Colunas: {list(df.columns)}")
            
            # Resultados de performance
            print(f"\n⏱️  PERFORMANCE:")
            print(f"   • Tempo de execução: {execution_time:.4f} segundos")
            print(f"   • Velocidade: {len(df)/execution_time:,.0f} linhas/segundo")
            
            print(f"\n🔋 RECURSOS DEPOIS - CPU: {resources_after['cpu_percent']:.1f}% | "
                  f"Memória: {resources_after['memory_mb']:.1f} MB ({resources_after['memory_percent']:.1f}%)")
            
            memory_diff = resources_after['memory_mb'] - resources_before['memory_mb']
            print(f"   • Diferença de memória: {memory_diff:+.1f} MB")
            
            # Salvar resultados
            result = {
                'file': os.path.basename(filepath),
                'method': method,
                'file_size_mb': file_size / 1024 / 1024,
                'rows': len(df),
                'execution_time': execution_time,
                'rows_per_second': len(df) / execution_time,
                'memory_before_mb': resources_before['memory_mb'],
                'memory_after_mb': resources_after['memory_mb'],
                'memory_diff_mb': memory_diff
            }
            
            self.results[f"{method}_{os.path.basename(filepath)}"] = result
            
            return df
            
        except Exception as e:
            print(f"❌ Erro ao ler arquivo: {e}")
            return None
    
    def basic_calculations(self, df):
        """
        Realiza cálculos básicos e mede performance
        """
        if df is None:
            print("❌ DataFrame vazio, pulando cálculos")
            return
        
        print(f"\n{'='*50}")
        print("EXECUTANDO CÁLCULOS BÁSICOS")
        print(f"{'='*50}")
        
        resources_before = self.measure_resources()
        start_time = time.time()
        
        try:
            # Cálculos básicos
            calculations = {
                'soma_total': df['value'].sum(),
                'media': df['value'].mean(),
                'mediana': df['value'].median(),
                'min_valor': df['value'].min(),
                'max_valor': df['value'].max(),
                'desvio_padrao': df['value'].std(),
                'count': len(df)
            }
            
            end_time = time.time()
            execution_time = end_time - start_time
            
            resources_after = self.measure_resources()
            memory_diff = resources_after['memory_mb'] - resources_before['memory_mb']
            
            print(f"\n📈 RESULTADOS DOS CÁLCULOS:")
            for key, value in calculations.items():
                if isinstance(value, float):
                    print(f"   • {key.replace('_', ' ').title()}: {value:,.2f}")
                else:
                    print(f"   • {key.replace('_', ' ').title()}: {value:,}")
            
            print(f"\n⏱️  PERFORMANCE CÁLCULOS:")
            print(f"   • Tempo: {execution_time:.4f} segundos")
            print(f"   • Memória extra: {memory_diff:+.1f} MB")
            
            return calculations
            
        except Exception as e:
            print(f"❌ Erro nos cálculos: {e}")
            return None
    
    def compare_reading_methods(self, filepath):
        """
        Compara diferentes métodos de leitura
        """
        print(f"\n{'='*60}")
        print("COMPARAÇÃO DE MÉTODOS DE LEITURA")
        print(f"{'='*60}")
        
        methods = ['pandas']  # Pode adicionar mais métodos depois
        
        for method in methods:
            df = self.read_csv_with_timing(filepath, method)
            if df is not None:
                self.basic_calculations(df)
        
        return self.results
    
    def print_summary(self):
        """
        Imprime resumo dos resultados
        """
        if not self.results:
            print("❌ Nenhum resultado para mostrar")
            return
        
        print(f"\n{'='*60}")
        print("RESUMO DOS RESULTADOS")
        print(f"{'='*60}")
        
        for key, result in self.results.items():
            print(f"\n📊 {key}:")
            print(f"   • Arquivo: {result['file']}")
            print(f"   • Tamanho: {result['file_size_mb']:.2f} MB")
            print(f"   • Linhas: {result['rows']:,}")
            print(f"   • Tempo: {result['execution_time']:.4f}s")
            print(f"   • Velocidade: {result['rows_per_second']:,.0f} linhas/s")
            print(f"   • Memória usada: {result['memory_diff_mb']:+.1f} MB")

def main():
    """
    Função principal
    """
    print("🚀 INICIANDO TESTE DE PERFORMANCE - PYTHON")
    print("="*60)
    
    processor = DataProcessor()
    
    # Caminhos dos arquivos
    large_dataset_path = '../data/large_dataset.csv'
    sample_dataset_path = '../data/sample_dataset.csv'
    
    # Verificar se o dataset grande existe
    if not os.path.exists(large_dataset_path):
        print(f"⚠️  Dataset grande não encontrado: {large_dataset_path}")
        print("🔄 Gerando dataset...")
        
        # Importar e executar gerador
        from generate_large_dataset import generate_large_dataset
        generate_large_dataset()
    
    # Testar com dataset grande
    if os.path.exists(large_dataset_path):
        print(f"\n🎯 TESTANDO COM DATASET GRANDE")
        processor.compare_reading_methods(large_dataset_path)
    
    # Testar com dataset pequeno para comparação
    if os.path.exists(sample_dataset_path):
        print(f"\n🎯 TESTANDO COM DATASET PEQUENO (comparação)")
        processor.compare_reading_methods(sample_dataset_path)
    
    # Mostrar resumo
    processor.print_summary()
    
    print(f"\n✅ TESTE CONCLUÍDO!")

if __name__ == "__main__":
    main()

File: go-vs-python-data-processing/go/process_data/process_data.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"math"
	"math/rand"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"time"
)

// DataRecord representa um registro de dados
type DataRecord struct {
	ID    int
	Value int
}

// DataProcessor gerencia o processamento de dados e medição de performance
type DataProcessor struct {
	results map[string]interface{}
}

// NewDataProcessor cria uma nova instância do processador
func NewDataProcessor() *DataProcessor {
	return &DataProcessor{
		results: make(map[string]interface{}),
	}
}

// MemoryStats obtém estatísticas de memória
func (dp *DataProcessor) GetMemoryStats() (float64, float64) {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)

	allocMB := float64(m.Alloc) / 1024 / 1024
	sysMB := float64(m.Sys) / 1024 / 1024

	return allocMB, sysMB
}

// ReadCSVWithTiming lê CSV e mede performance
func (dp *DataProcessor) ReadCSVWithTiming(filepath string) ([]DataRecord, error) {
	fmt.Printf("\n%s\n", strings.Repeat("=", 50))
	fmt.Printf("LENDO CSV: %s\n", filepath)
	fmt.Printf("Método: Go encoding/csv\n")
	fmt.Printf("%s\n", strings.Repeat("=", 50))

	// Verificar se arquivo existe
	fileInfo, err := os.Stat(filepath)
	if err != nil {
		fmt.Printf("❌ Arquivo não encontrado: %s\n", filepath)
		return nil, err
	}

	fileSize := fileInfo.Size()
	fmt.Printf("📁 Tamanho do arquivo: %d bytes (%.2f MB)\n", fileSize, float64(fileSize)/(1024*1024))

	// Medição de recursos antes
	allocBefore, sysBefore := dp.GetMemoryStats()
	fmt.Printf("🔋 Recursos ANTES - Alloc: %.1f MB | Sys: %.1f MB\n", allocBefore, sysBefore)

	// Medir tempo de leitura
	startTime := time.Now()

	// Abrir arquivo
	file, err := os.Open(filepath)
	if err != nil {
		return nil, fmt.Errorf("erro ao abrir arquivo: %v", err)
	}
	defer file.Close()

	// Criar reader CSV
	reader := csv.NewReader(file)

	// Ler cabeçalho
	headers, err := reader.Read()
	if err != nil {
		return nil, fmt.Errorf("erro ao ler cabeçalho: %v", err)
	}
	_ = headers // Ignorar cabeçalho por enquanto

	// Ler dados
	var records []DataRecord
	lineCount := 0

	for {
		record, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("erro ao ler linha %d: %v", lineCount+1, err)
		}

		// Converter dados
		id, err := strconv.Atoi(record[0])
		if err != nil {
			return nil, fmt.Errorf("erro ao converter ID na linha %d: %v", lineCount+1, err)
		}

		value, err := strconv.Atoi(record[1])
		if err != nil {
			return nil, fmt.Errorf("erro ao converter Value na linha %d: %v", lineCount+1, err)
		}

		records = append(records, DataRecord{ID: id, Value: value})
		lineCount++
	}

	executionTime := time.Since(startTime)

	// Medição de recursos depois
	allocAfter, sysAfter := dp.GetMemoryStats()

	// Forçar garbage collection para medição mais precisa
	runtime.GC()

	fmt.Printf("\n📊 DADOS CARREGADOS:\n")
	fmt.Printf("   • Linhas: %d\n", len(records))
	fmt.Printf("   • Colunas: 2 (id, value)\n")

	fmt.Printf("\n⏱️  PERFORMANCE:\n")
	fmt.Printf("   • Tempo de execução: %.4f segundos\n", executionTime.Seconds())
	fmt.Printf("   • Velocidade: %.0f linhas/segundo\n", float64(len(records))/executionTime.Seconds())

	memoryDiff := allocAfter - allocBefore
	fmt.Printf("\n🔋 RECURSOS DEPOIS - Alloc: %.1f MB | Sys: %.1f MB\n", allocAfter, sysAfter)
	fmt.Printf("   • Diferença de memória: %+.1f MB\n", memoryDiff)

	return records, nil
}

// CalculateStatistics calcula estatísticas básicas
func (dp *DataProcessor) CalculateStatistics(records []DataRecord) map[string]float64 {
	if len(records) == 0 {
		fmt.Println("❌ Sem dados para calcular")
		return nil
	}

	fmt.Printf("\n%s\n", strings.Repeat("=", 50))
	fmt.Printf("EXECUTANDO CÁLCULOS BÁSICOS\n")
	fmt.Printf("%s\n", strings.Repeat("=", 50))

	allocBefore, _ := dp.GetMemoryStats()
	startTime := time.Now()

	// Extrair valores para cálculos
	values := make([]int, len(records))
	for i, record := range records {
		values[i] = record.Value
	}

	// Cálculos básicos
	sum := 0
	min := values[0]
	max := values[0]

	for _, value := range values {
		sum += value
		if value < min {
			min = value
		}
		if value > max {
			max = value
		}
	}

	mean := float64(sum) / float64(len(values))

	// Mediana
	sortedValues := make([]int, len(values))
	copy(sortedValues, values)
	sort.Ints(sortedValues)

	var median float64
	n := len(sortedValues)
	if n%2 == 0 {
		median = float64(sortedValues[n/2-1]+sortedValues[n/2]) / 2
	} else {
		median = float64(sortedValues[n/2])
	}

	// Desvio padrão
	var variance float64
	for _, value := range values {
		variance += math.Pow(float64(value)-mean, 2)
	}
	variance /= float64(len(values))
	stdDev := math.Sqrt(variance)

	// Quartis
	q1 := float64(sortedValues[n/4])
	q3 := float64(sortedValues[3*n/4])

	// Valores únicos
	uniqueMap := make(map[int]bool)
	for _, value := range values {
		uniqueMap[value] = true
	}
	uniqueCount := len(uniqueMap)

	calculations := map[string]float64{
		"soma_total":    float64(sum),
		"media":         mean,
		"mediana":       median,
		"min_valor":     float64(min),
		"max_valor":     float64(max),
		"desvio_padrao": stdDev,
		"quartil_25":    q1,
		"quartil_75":    q3,
		"count":         float64(len(values)),
		"unique_count":  float64(uniqueCount),
	}

	executionTime := time.Since(startTime)
	allocAfter, _ := dp.GetMemoryStats()
	memoryDiff := allocAfter - allocBefore

	fmt.Printf("\n📈 RESULTADOS DOS CÁLCULOS:\n")
	for key, value := range calculations {
		if key == "count" || key == "unique_count" {
			fmt.Printf("   • %s: %.0f\n", strings.Title(strings.ReplaceAll(key, "_", " ")), value)
		} else {
			fmt.Printf("   • %s: %.2f\n", strings.Title(strings.ReplaceAll(key, "_", " ")), value)
		}
	}

	fmt.Printf("\n⏱️  PERFORMANCE CÁLCULOS:\n")
	fmt.Printf("   • Tempo: %.4f segundos\n", executionTime.Seconds())
	fmt.Printf("   • Memória extra: %+.1f MB\n", memoryDiff)

	return calculations
}

// CompareReadingMethods compara diferentes métodos (por enquanto só um)
func (dp *DataProcessor) CompareReadingMethods(filepath string) {
	fmt.Printf("\n%s\n", strings.Repeat("=", 60))
	fmt.Printf("COMPARAÇÃO DE MÉTODOS DE LEITURA\n")
	fmt.Printf("%s\n", strings.Repeat("=", 60))

	records, err := dp.ReadCSVWithTiming(filepath)
	if err != nil {
		fmt.Printf("❌ Erro: %v\n", err)
		return
	}

	if records != nil {
		dp.CalculateStatistics(records)
	}
}

// PrintSummary imprime resumo dos resultados
func (dp *DataProcessor) PrintSummary() {
	fmt.Printf("\n%s\n", strings.Repeat("=", 60))
	fmt.Printf("RESUMO DOS RESULTADOS\n")
	fmt.Printf("%s\n", strings.Repeat("=", 60))

	// Por enquanto, apenas mensagem informativa
	fmt.Printf("Processamento concluído com sucesso!\n")
}

// GenerateDataset gera um CSV com numRows linhas
func GenerateDataset(numRows int, filename string) (string, error) {
	dataDir := "../data"
	if err := os.MkdirAll(dataDir, 0755); err != nil {
		return "", fmt.Errorf("erro ao criar diretório: %v", err)
	}

	filePath := filepath.Join(dataDir, filename)

	file, err := os.Create(filePath)
	if err != nil {
		return "", fmt.Errorf("erro ao criar arquivo: %v", err)
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	if err := writer.Write([]string{"id", "value"}); err != nil {
		return "", fmt.Errorf("erro ao escrever cabeçalho: %v", err)
	}

	rand.Seed(time.Now().UnixNano())
	for i := 1; i <= numRows; i++ {
		value := rand.Intn(4950) + 50
		record := []string{fmt.Sprintf("%d", i), fmt.Sprintf("%d", value)}
		if err := writer.Write(record); err != nil {
			return "", fmt.Errorf("erro ao escrever linha %d: %v", i, err)
		}
	}

	return filePath, nil
}

func main() {
	fmt.Printf("🚀 INICIANDO TESTE DE PERFORMANCE - GO\n")
	fmt.Printf("%s\n", strings.Repeat("=", 60))

	processor := NewDataProcessor()

	// Caminhos dos arquivos
	largeDatasetPath := "../data/large_dataset_go.csv"
	sampleDatasetPath := "../data/sample_dataset.csv"

	// Verificar se o dataset grande existe
	if _, err := os.Stat(largeDatasetPath); os.IsNotExist(err) {
		fmt.Printf("⚠️  Dataset grande não encontrado: %s\n", largeDatasetPath)
		fmt.Printf("🔄 Gerando dataset...\n")

		if newPath, err := GenerateDataset(10000, "large_dataset_go.csv"); err != nil {
			fmt.Printf("❌ Erro ao gerar dataset: %v\n", err)
			return
		} else {
			largeDatasetPath = newPath
		}
	}

	// Testar com dataset grande
	if _, err := os.Stat(largeDatasetPath); err == nil {
		fmt.Printf("\n🎯 TESTANDO COM DATASET GRANDE\n")
		processor.CompareReadingMethods(largeDatasetPath)
	}

	// Testar com dataset pequeno para comparação
	if _, err := os.Stat(sampleDatasetPath); err == nil {
		fmt.Printf("\n🎯 TESTANDO COM DATASET PEQUENO (comparação)\n")
		processor.CompareReadingMethods(sampleDatasetPath)
	}

	// Mostrar resumo
	processor.PrintSummary()

	fmt.Printf("\n✅ TESTE CONCLUÍDO!\n")
}
